### PR TITLE
Updates to generalized architecture

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
Updated the generalized architecture merging the work from the vendor we hired. There's been quite a bit of cleanup done, simplifying the architecture.

Current status:

1. The template works. It deploys, and the cluster comes up and working. I was able to confirm it's working with every pre-made "t-shirt size". Haven't fully tested all the possible configurations, however.
2. The documentation hasn't been fully updated yet. The README is up to date and has been expanded. In the docs folder, some pages need extra work - I'll be working on that in the next days and update the PR.
3. I haven't performed load tests yet.

### CODE REVIEW

File: docs/Cleanup.md [hunk 1]
```diff
@@ -10,12 +10,9 @@ First we need to ensure our [environment variables](./Environment-Variables.md)
...
-with caution.
+This command will delete all resources in *all* resource groups. Run with caution.
 
-Note, that this command will not fully delete the resource group if
-you have Azure Backup enabled since the Recovery Services Vault will
-not be deleted (it's got the backups of you data!).
+Note, that this command will not fully delete the resource group if you have Azure Backup enabled since the Recovery Services Vault will not be deleted (it's got the backups of you data!).
 
 ``` bash
 for filename in $MOODLE_AZURE_WORKSPACE/*; do az group delete --yes --name $(basename $filename) --no-wait; done
```
Comment (rgardler) [inline docs/Cleanup.md]: Still got a Moodle reference here (and likely in other places where this environment variables are used)

File: docs/Manage.md [hunk 2]
```diff
@@ -71,7 +71,7 @@ web VMs through the specified file server (Gluster or NFS), and this is
 not good for web response time. Therefore, we introduced the
 `htmlLocalCopySwitch` that'll copy the `/azlamp/html` directory to
 `/var/www/html` in each autoscaled web VM and reconfigures the web
-server (apache/nginx)'s server root directory accordingly, when it's set
+server (nginx)'s server root directory accordingly, when it's set
```
Comment (rgardler) [inline docs/Manage.md]: Has Apache been removed? Why? Many customers prefer Apache, especially for PHP workloads.

File: docs/Parameters.md [hunk 3]
```diff
@@ -31,7 +31,7 @@ Type: string
 
 Possible Values: null
 
-Default: https://raw.githubusercontent.com/Azure/Moodle/master/
+Default: https://raw.githubusercontent.com/Azure/LAMP/master/
```
Comment (rgardler) [inline docs/Parameters.md]: This file should not be manually edited, see notes at top of file. If this was generated by the script then all good.

File: docs/Readme.md [hunk 4]
```diff
@@ -31,7 +31,7 @@ NOTE: Depending on the region you choose to deploy the stack in - the deployment
 ## Stack Architecture
 
 This template set deploys the following infrastructure core to your LAMP instance:
-- Autoscaling web frontend layer (Nginx for https termination, Varnish for caching, Apache/php or nginx/php-fpm)
+- Autoscaling web frontend layer (Nginx for https termination, Apache/php or nginx/php-fpm)
```
Comment (rgardler) [inline docs/Readme.md]: Has Varnish been removed? If so why? It was added in for a specific reason (though I do not recall why). Has load testing been done against this configuration?

**Comment by ItalyPaleAle** [comment]
Hey @rgardler thanks for the review:

1. Deployments are working (functionally working, by installing both WordPress and another app on the same cluster, not just "it deploys"), but haven't run load tests yet.
1. The docs in the docs/ folder haven't been fully updated yet. Got it on the file that needs to be auto-generated, I'll fix that.
1. The configuration for Varnish is very app-dependent. The scripts before were specific to Moodle (see for example: https://github.com/Azure/LAMP/blob/master/scripts/helper_functions.sh#L1068 ). When I spoke with the vendor, we felt that we could accomplish a better result for a generic LAMP app, in a much simpler way, by recommending the use of an external CDN rather than adding "our own CDN".
1. In the current version of the code, there were both Apache2 and Nginx configured. Apache2 was serving only HTTP requests, while Nginx was serving only HTTPS requests. We were not sure what the reason was, probably something Moodle-specific? If you prefer to just use Apache2 rather than Nginx, I'll switch back

**Comment by rgardler** [comment]
1 - 3 sounds good

Item 4:
Apache was handling all the PHP requests, NGinx was merely a termination for HTTPS requests. It did not serve content.

The original templates only used NGinx but we heard that many customers preferred the debug experience in Apache. The Apache + NGinx solution was always a bit of a cludge.

Don't put it back but be aware of this and listen for user feedback.

LGTM.

**Comment by ItalyPaleAle** [comment]
I looked into it more.

I don't know the rationale for picking Apache2 as main webserver, but looking at the code again now I think it *might* have been because of the .htaccess and mod_rewrite support? On Nginx you can achieve those things by changing the website configuration in the /etc folder. However, I noticed now how the Nginx config is created on-the-go for each VMSS node. So, I think there could be two options here:

1. Revert back to use Apache2, so .htaccess files would work and people could easily configure URL rewrites. In this case, however, would it make sense to terminate SSL directly on Apache, rather than having another server running?
2. Move the Nginx configuration to the /azlamp folder for persistent storage and generate it on the controller (then replicate on each node), so users could modify them once for all.

What do you think?

**Comment by rgardler** [comment]
See my overlapping comment which gives the reason

**Comment by ItalyPaleAle** [comment]
I've updated the documentation. Please let me know if this is good to merge (I will still make changes after that)

**Comment by rgardler** [comment]
I have *not* tested this, but it LGTM. If you are confident in its stability go ahead and merge.

**Pull request merged**